refactor: Fetch and cache GutenbergKit editor settings

dcalhoun opened the pull request:
Fetch and cache editor settings before providing them to the editor, as it
allows improving offline support. The changes also communicate the 
background activity with an activity indicator. 

Ref CMM-200.

Related:

* CMM-200-linear-issue
* https://github.com/wordpress-mobile/GutenbergKit/issues/105
* https://github.com/wordpress-mobile/GutenbergKit/pull/114
* https://github.com/wordpress-mobile/WordPress-iOS/pull/24417

## To Test:
<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

> [!Important]
> Ensure the following experimental features are enabled:
> * Experimental block editor
> * Experimental block editor styles

**1: Initial editor fallbacks to default settings**
1. Enable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads, using default settings—e.g., white background with black text. 

**2: Editor fetches site-specific settings**
1. Disable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads with site-specific settings—e.g., the theme's background and foreground colors. 

**3: Editor fallback to cached settings**
1. Enable airplane mode. 
2. Open the post editor. 
3. Verify the editor loads with site-specific settings—e.g., the theme's background and foreground colors. 

## Regression Notes

1. Potential unintended areas of impact
  Regressions in Gutenberg Mobile or Aztec editors. 
1. What I did to test those areas of impact (or what existing automated tests I relied on)
  Manually smoke tested the editors. 
1. What automated tests I added (or what prevented me from doing so)
  Deemed unnecessary for the experimental editor. 

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

dcalhoun committed:
refactor: Provide editor settings to GutenbergKit

dcalhoun committed:
refactor: Move onEditorSettingsChanged subscription to EditPostActivity

Mirror theme change subscription.

dcalhoun committed:
style: Automated formatting changes

dcalhoun committed:
feat: Dispatch editor settings request

dcalhoun committed:
refactor: Improve editor settings parsing

dcalhoun committed:
refactor: Simplify EditorSettings

dcalhoun committed:
refactor: Pass editor settings as JSON string

Avoid typing complex JSON structures.

dcalhoun committed:
feat: Display activity indicator while fetching editor settings

Communicate active background work to the user.

dcalhoun committed:
feat: Add cache-while-revalidate for editor settings

Improve offline support and performance.

dcalhoun committed:
feat: Persist editor settings in SQLite

Match caching strategy for other FluxC stores.

dcalhoun committed:
build: Update GutenbergKit ref

dangermattic commented:
<!--
  7 Errors: Please add tests for class <co...
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          7 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettingsAction</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettings</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettingsSqlUtils</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettingsBuilder</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettingsStore</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>FetchEditorSettingsPayload</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EditorSettingsError</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
style: Address lint errors

dcalhoun committed:
style: Address lint warnings

dcalhoun committed:
fix: Use -1 post ID for new drafts

This is the expected ID value for the Gutenberg editor.

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:
```diff
@@ -3633,9 +3639,12 @@ class EditPostActivity : BaseAppCompatActivity(), EditorFragmentActivity, Editor
 
     private fun onEditorFinalTouchesBeforeShowing() {
-        refreshEditorContent()
+        if (editorFragment !is GutenbergKitEditorFragment) {
+            refreshEditorContent()
+        }
```
The timing of this content refresh is currently incompatible with/unnecessary for GutenbergKit.

dcalhoun committed:
Revert "style: Automated formatting changes"

This reverts commit a71542448ae3ee7a4eeb2ba1668a1e0e86e860e1.

dcalhoun commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt:
```diff
@@ -41,7 +41,7 @@ open class WellSqlConfig : DefaultWellConfig {
     annotation class AddOn
 
     override fun getDbVersion(): Int {
-        return 204
```
Noting that a 204 migration did not exist. I do not believe this is a problem, but noting it just in case.

dcalhoun commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/EditorSettingsStore.kt:
This store was built based on the existing `EditorThemeStore`.

nbradbury replied:
@dcalhoun I'm curious why we have this [duplication](https://github.com/wordpress-mobile/WordPress-Android/blob/f12ea63efc0ce5c248d0e00438d9e96254fecb95/libs/fluxc/src/main/java/org/wordpress/android/fluxc/store/EditorThemeStore.kt#L130) with `EditorThemeStore`? Do we need two separate stores for editor settings?

dcalhoun replied:
@nbradbury it's an attempt to mitigate regressions. 

When the GutenbergKit implementation first began, we colocated much of the new logic with existing editor logic; this included introducing a good amount of branching logic to exiting code. In retrospect, we deemed this misguided, as every change we made in the advancement of GutenbergKit meant the possibility of a regression in the Gutenberg Mobile and Aztec editors. We ultimately separated a good amount of the logic in https://github.com/wordpress-mobile/WordPress-Android/pull/21493. 

Duplicating `EditorThemeStore` to `EditorSettings` felt appropriate for continuing that separation practice, it also allowed for a simpler implementation for GutenbergKit.

dcalhoun committed:
refactor: Remove unnecessary style configuration

This was added erroneously.

dcalhoun committed:
refactor: Avoid redundant editor starts

Now that we await the editor settings, we should postpone starting the
editor. This means we only start the editor once.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
style: Address lint warning

dcalhoun committed:
fix: Start editor with default settings on failed editor settings

Ensure the editor continues launching if retrieving site-specific editor
settings fails.

dcalhoun committed:
fix: Emit available cached editor settings on fetch failure

Apply the cached settings rather than the default settings, if
available.

nbradbury commented:
@dcalhoun This isn't something that needs to be addressed here, especially since `EditorThemeStore` already does this, but I wanted to caution against storing raw JSON in the database. I think it was with WCAndroid where we encountered users getting [SQLiteBlobTooBigException](https://developer.android.com/reference/android/database/sqlite/SQLiteBlobTooBigException) due to plugins adding a ton of baggage to JSON responses.

Semi-related, we may want to make use of the `fields` parameter to limit the response to just the fields we use. The response is only ~5KB in this case so it's not huge, but it would be more efficient for storage. 

![response](https://github.com/user-attachments/assets/00f44847-7025-4978-b4ef-4a66d94af7b6)

dcalhoun commented:
> This isn't something that needs to be addressed here, especially since `EditorThemeStore` already does this, but I wanted to caution against storing raw JSON in the database. I think it was with WCAndroid where we encountered users getting [SQLiteBlobTooBigException](https://developer.android.com/reference/android/database/sqlite/SQLiteBlobTooBigException) due to plugins adding a ton of baggage to JSON responses.

@nbradbury thank you for the feedback. This is a great point. I believe the WP-Android app suffers from similar exceptions in certain contexts. 

> Semi-related, we may want to make use of the `fields` parameter to limit the response to just the fields we use. The response is only ~5KB in this case so it's not huge, but it would be more efficient for storage.

Leveraging the `_fields` parameter may be a good option. I have not reviewed which fields are applicable to GutenbergKit at this time; doing so is a good optimization opportunity to pursue later. 

Is there an alternative persistence layer to SQLite that makes sense for this specific feature? Or is SQLite the best choice? For context, the data is site-specific and should likely persist _at least_ a few days.

nbradbury commented:
> Is there an alternative persistence layer to SQLite that makes sense for this specific feature? Or is SQLite the best choice?

I recommend sticking with SQLite, especially for this feature since the data isn't unreasonably large. I've heard other devs recommend storing the data in a separate file and write the filename to the database, but I've never felt the need to do that.

nbradbury reviewed (approved):
Looks good! :shipit:

dcalhoun commented:
@nbradbury thank you for the review. Did you have time to also review the sibling PR: https://github.com/wordpress-mobile/GutenbergKit/pull/114?

dcalhoun committed:
build: Update GutenbergKit ref
